rules/firestore.json

```diff
@@ -35,50 +35,205 @@ service cloud.firestore {
+            // A part is started when the beginning field is set
+            return hasField(part, 'beginning');
+        }
+        function hasTimedOut(user) {
+            let userDoc =  get(/databases/$(database)/documents/users/$(user.id)).data;
+            let timedOut = duration.value(request.time.toMillis(), 'ms') > duration.value(userDoc.last_changed.seconds, 's') + duration.value(10, 's');
+            return timedOut;
+        }
+        match /joiners/{partId} {
+            // Any verified user can create a joiner as long as they are themselves creator and if the corresponding part exists
+            allow create: if userIsVerified() &&
+                             userIsCreatorWithValidName(request.resource.data) &&
+                             exists(/databases/$(database)/documents/parts/$(partId));
+            // The creator can delete a joiner if the part has not started
+            allow delete: if userIsVerified() && userIsCreator(resource.data) && joinerHasNotStarted(resource.data);
+            // Anybody else can delete a joiner only if the creator has timed out and the part has not started
+            allow delete: if userIsVerified() && hasTimedOut(creatorOf(partId)) && joinerHasNotStarted(resource.data);
+            // The creator can change all fields except changing partStatus to STARTED (3), until the part starts
+            allow update: if userIsVerified() && userIsCreator(resource.data) &&
+                             resource.data.partStatus < 3 && request.resource.data.partStatus != 3;
+            // The creator can only change partStatus (to finished)  once the part has been started
+            allow update: if userIsVerified() && userIsCreator(resource.data) &&
+                             resource.data.partStatus == 3 &&
+                             (request.resource.data.partStatus >= 3) &&
+                             preservesField('creator') &&
+                             preservesField('firstPlayer') &&
+                             preservesField('partType') &&
+                             preservesField('maximalMoveDuration') &&
+                             preservesField('totalPartDuration') &&
+                             preservesField('chosenOpponent')
+            // For anybody else:
+            //   - the chosen opponent can change the config from PROPOSED to STARTED
+            allow update: if validJoinerUpdate() &&
+                             preservesField('chosenOpponent') &&
+                             acceptsConfig()
+            //   - the chosen opponent can remove itself from the corresponding field
+            allow update: if validJoinerUpdate() &&
+                             preservesField('partStatus') &&
+                             removesOneSelfFromChosenOpponent();
+            // Any verified user can read a joiner
+            allow read: if userIsVerified();
+        }
+        function validJoinerUpdate() {
```

Comment:
Donc ici par Joiner on entend, le type de document ?
Une modification du partType est pourtant légale, et le créateur peut aussi modifier le document ?

Reply:
Je suis pressé qu'on dise plus Joiner pour le type de document mais ça ne sera acceptable que quand celui ci sera appellé ConfigRoom :D !

Reply:
désolé, three-in-a-row: mais du coup si faut être chosenCandidate pour que ce soit accepté, c'est validChosenCandidateUpdate et ça résoulde les deux précédenr !

Reply:
NON :D C'est `validChosenOpponentUpdate` désormais !
Et oui pour joiner -> ConfigRoom ! J'renomme dans un commit séparé par après (TODO)